rust/consensus/src/lib.rs

```diff
@@ -1,404 +1,301 @@
-                        return support_stake;
-                    }
-                }
+    /// Order the past leaders that we didn't already commit.
+    fn order_leaders(&self, leader: &Certificate, state: &State) -> Vec<Certificate> {
+        let mut to_commit = vec![leader.clone()];
+        let mut leader = leader;
+        for r in (state.last_committed_round + 2..leader.round())
+            .rev()
+            .step_by(2)
+        {
+            // Get the certificate proposed by the previous leader.
+            let (_, prev_leader) = match self.leader(r, &state.dag) {
+                Some(x) => x,
+                None => continue,
+            };
+
+            // Check whether there is a path between the last two leaders.
+            if self.linked(leader, prev_leader, &state.dag) {
+                to_commit.push(prev_leader.clone());
+                leader = prev_leader;
             }
         }
-        0
+        to_commit
     }
 
-    pub fn compute_weak_support(&self, round: RoundNumber, leader: NodeID) -> Stake {
-        //first check that you see the leaders block
-        if let Some((leader_digest, _)) = self.get_header_digest_by_node_round(leader, round - 1) {
-            //Then get all its direct ancestors
-            if let Some(headers) = self.get_headers(round) {
-                let support_stake: Stake = headers
-                    .iter()
-                    .filter(|(_, header)| header.has_parent(leader_digest))
-                    .map(|(_, header)| self.committee.stake(&header.author))
-                    .sum();
-                //this is the support
-                return support_stake;
-            }
+    /// Checks if there is a path between two leaders.
+    fn linked(&self, leader: &Certificate, prev_leader: &Certificate, dag: &Dag) -> bool {
+        let mut parents = vec![leader];
+        for r in (prev_leader.round()..leader.round()).rev() {
+            parents = dag
+                .get(&(r))
+                .expect("We should have the whole history by now")
+                .values()
+                .filter(|(digest, _)| parents.iter().any(|x| x.header.parents.contains(digest)))
+                .map(|(_, certificate)| certificate)
+                .collect();
         }
-        0
+        parents.contains(&prev_leader)
     }
 
-    pub fn get_header_digest_by_node_round(
-        &self,
-        node: NodeID,
-        round: RoundNumber,
-    ) -> Option<(&Digest, &BlockHeader)> {
-        if let Some(headers) = self.headers_dag.get(&round) {
-            for (digest, header) in headers {
-                if header.author == node {
-                    return Some((digest, header));
+    /// Flatten the dag referenced by the input certificate. This is a classic depth-first search (pre-order):
+    /// https://en.wikipedia.org/wiki/Tree_traversal#Pre-order
+    fn order_dag(&self, leader: &Certificate, state: &State) -> Vec<Certificate> {
```

Comment:
you should stop at last_committed-gc_depth the DAG otherwise you will hit garbage collected state

Reply:
I think we do stop there: https://github.com/facebookresearch/narwhal/blob/150a46aa587eb4a72abf9d7812e36a54470e7daa/rust/consensus/src/lib.rs#L279